### PR TITLE
Update JSON Schema documentation for V2

### PR DESCRIPTION
**Pull request by tpdorsey**
## Related issue number

Closes PYD-116

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @adriangb

### CODE REVIEW

File: docs/usage/json_schema.md [hunk 1]
```diff
@@ -826,4 +715,15 @@ print(json.dumps(Person.model_json_schema(), indent=2))
...
+## Customizing the JSON schema generation process
+
+If you need custom schema generation, you can use a `schema_generator`, modifying the
+[`GenerateJsonSchema`][pydantic.json_schema.GenerateJsonSchema] class as necessary for your application.
+
+The various methods that can be used to produce JSON schema accept a keyword argument `schema_generator: type[GenerateJsonSchema] = GenerateJsonSchema`, and you can pass your custom subclass to these methods in order to use your own approach to generating JSON schema.
+
+`GenerateJsonSchema` implements the translation of a type's `pydantic-core` schema into a JSON schema.
+By design, this class breaks the JSON schema generation process into smaller methods that can be easily overridden in
+subclasses to modify the "global" approach to generating JSON schema.
```
Comment (Kludex) [inline docs/usage/json_schema.md:729]: I think there's an example that we can use from our code source with discriminators that we can give as an example here.

**Comment by tpdorsey** [comment]
please review

File: docs/usage/json_schema.md [hunk 2]
```diff
@@ -1,7 +1,20 @@
...
+Pydantic provides several ways to generate JSON schemas or JSON representations from fields or models:
+
+* [`BaseModel.model_json_schema`][pydantic.main.BaseModel.model_json_schema] returns a dict of the schema.
+* [`BaseModel.model_dump_json`][pydantic.main.BaseModel.model_dump_json] returns a JSON string representation of the
+    dict of the schema.
+* [`TypeAdapter.dump_json`][pydantic.type_adapter.TypeAdapter.dump_json] serializes an instance of the adapted type to
+    JSON.
+* [`TypeAdapter.json_schema`][pydantic.type_adapter.TypeAdapter.json_schema] generates a JSON schema for the adapted type.
+* [`TypeAdapter.json_schemas`][pydantic.type_adapter.TypeAdapter.json_schemas] generates a JSON schema including
+    definitions from multiple type adapters.
```
Comment (adriangb) [inline docs/usage/json_schema.md:12]: Can we skip documenting this? I think we may want to remove it...

File: docs/usage/json_schema.md [hunk 3]
```diff
@@ -826,4 +716,67 @@ print(json.dumps(Person.model_json_schema(), indent=2))
 """
 ```
 
-Note that you *must* return a schema, even if you are just mutating it in-place.
+Note that you *must* return a schema, even if you are just mutating it in place.
+
+## Customizing the JSON schema generation process
+
+If you need custom schema generation, you can use a `schema_generator`, modifying the
+[`GenerateJsonSchema`][pydantic.json_schema.GenerateJsonSchema] class as necessary for your application.
```
Comment (adriangb) [inline docs/usage/json_schema.md:724]: @dmontagu I defer to you if you want to document this, or document it with a big warning at the top saying it's experimental?

**Review by adriangb — APPROVED** [review]
left some notes but generally looks good!

File: docs/usage/json_schema.md (hunk 3 above)
Comment (dmontagu) [reply docs/usage/json_schema.md:721]: I think let's  hold off on documenting for now but plan to include it with a warning when we add it, maybe shortly. I'd like to give FastAPI a bit of time first.

**Pull request merged**